addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/storage-gateway/create-network-infrastructure.js

```diff
@@ -0,0 +1,254 @@
+    };
+    const gatewayARN = await storageGateway.activateGateway(requestContext, rawData);
+    this.print(`Found GatewayARN as ${gatewayARN} from storage-gateway`);
+    this.state.setKey('GATEWAY_ARN', gatewayARN);
+  }
+
+  async validateGateway() {
+    // Validate that the gateway creation went through. Without this validation we cannot perform any operations
+    // on the Gateway like adding cache
+    const storageGateway = await this.getStorageGateway();
+    const requestContext = await this.payload.object('requestContext');
+    try {
+      const gatewayARN = await this.state.string('GATEWAY_ARN');
+      await storageGateway.listLocalDisks(requestContext, gatewayARN);
+    } catch (error) {
+      this.print(`Gateway validation failed: ${error.stack}`);
+      return false;
+    }
+    return true;
+  }
+
+  async addCacheToGateway() {
+    // Add the volume created by this workflow step as a cache
+    const gatewayARN = await this.state.string('GATEWAY_ARN');
+    const volumeId = await this.state.string('VOLUME_ID');
+    const rawData = {
+      gatewayARN,
+      volumeId,
+    };
+    const requestContext = await this.payload.object('requestContext');
+    const storageGateway = await this.getStorageGateway();
+    await storageGateway.addCacheToGateway(requestContext, rawData);
+
+    // save result to DDB
+    const vpcId = await this.state.string('VPC_ID');
+    const subnetId = await this.state.string('SUBNET_ID');
+    const ec2Instance = await this.state.string('EC2_INSTANCE_ID');
+    const elasticIP = await this.state.string('PUBLIC_IP');
+    const securityGroup = await this.state.string('SECURITY_GROUP');
+    const ec2RoleARN = await this.state.string('IAM_ROLE');
+    const dbData = {
+      vpcId,
+      subnetId,
+      ec2Instance,
+      elasticIP,
+      securityGroup,
+      ec2RoleARN,
+      volumeIds: [volumeId],
+    };
+    await storageGateway.saveToDDB(requestContext, dbData, gatewayARN);
+  }
+
+  async onFail() {
+    // 1. Delete the stack if it was created
+    // 2. Delete the Storage Gateway if it was created
+    try {
+      const stackId = await this.state.string('STATE_STACK_ID');
+      const cfn = await this.getCloudFormationService();
+      const stackInfo = (await cfn.describeStacks({ StackName: stackId }).promise()).Stacks[0];
+      this.print(`Found stack status: ${stackInfo.StackStatus}`);
+      if (stackInfo.StackStatus !== 'DELETE_COMPLETE') {
+        this.print(`Deleting stack: ${stackId}`);
+        await cfn.deleteStack({ StackName: stackId }).promise();
+        this.print(`Deleted stack: ${stackId}`);
+      }
+    } finally {
+      const storageGateway = await this.getStorageGateway();
+      const requestContext = await this.payload.object('requestContext');
+      const gatewayARN = await this.state.optionalString('GATEWAY_ARN');
+      if (gatewayARN) {
+        await storageGateway.deleteGateway(requestContext, gatewayARN);
```

Comment:
Will this delete all the resources that might be created by the stack?

Reply:
Before this line we are deleting the stack. But your question brought an interesting point, so I am going to store the CloudFormation stackId in DDB as well.